src/js/tls.js

```diff
@@ -257,8 +241,14 @@ function connect(arg0, arg1, arg2, callback) {
       callback = arg1;
     }
   }
-  tlsSocket = new TLSSocket(new net.Socket(), options);
-  tlsSocket.connect(options, callback);
+
+  var tlsSocket = new TLSSocket(options.socket || new net.Socket(), options);
+  if (tlsSocket._socket instanceof net.Socket) {
+    tlsSocket.connect(options, callback);
+  }
+  if (util.isFunction(callback)) {
```

Comment:
Isn't this should be an `else if`?

Reply:
I just realized you're right, yes, it has to be else if. The way I did it caused https tests to fail or timeout.